frontend/cachesqlite.lua

```diff
@@ -0,0 +1,178 @@
+            key TEXT PRIMARY KEY,
+            value BLOB,
+            size INTEGER,
+            last_access INTEGER
+        );
+        CREATE INDEX IF NOT EXISTS idx_last_access ON cache(last_access);
+    ]]
+
+    -- Initialize size tracking
+    self.current_size = self.db:rowexec([[
+        SELECT
+            ROUND(page_count * page_size, 0) as db_size_bytes,
+            ROUND((page_count - freelist_count) * page_size, 0) as data_size_bytes
+        FROM pragma_page_count(),
+            pragma_page_size(),
+            pragma_freelist_count();
+    ]]) or 0
+    logger.dbg("CacheSQLite:", self.db_path, "current size:", self.current_size)
+    self:closeDB()
+
+    self._persist = Persist.getCodec(self.codec)
+end
+
+--- Opens the SQLite database.
+--- This is normally done internally, but can be called manually if needed.
+function CacheSQLite:openDB()
+    self.db = SQ3.open(self.db_path)
+end
+
+--- Closes the SQLite database.
+--- This is normally done internally, but can be called manually if needed.
+function CacheSQLite:closeDB()
+    self.db:close()
+end
+
+--- Inserts an object into the cache.
+--- @param key string
+--- @param object any
+--- @return boolean success, number size
+function CacheSQLite:insert(key, object)
+    self:openDB()
+    local codec = Persist.getCodec(self.codec)
+    local size
+    object, size = codec.serialize(object)
+    if type(size) == "cdata" then
+        size = tonumber(size)
+    end
+    if not size then
+        size = #object
+    end
+
+    if not self:willAccept(size) then
+        logger.warn("Too much memory would be claimed by caching", key)
+        return false, 0
+    end
+
+    -- Ensure we have enough space by removing old entries
+    while self.current_size + size > self.size do
+        local oldest = self.db:rowexec([[
+            SELECT key, size FROM cache
+            ORDER BY last_access ASC LIMIT 1
+        ]])
+        if oldest then
+            self:remove(oldest[1])
+        else
+            break
+        end
+    end
+
+    -- Insert new entry using prepared statement
+    local stmt = self.db:prepare([[
+        INSERT OR REPLACE INTO cache (key, value, size, last_access)
+        VALUES (?, ?, ?, ?);
+    ]])
+    local ok, err = pcall(function()
+        stmt:reset():bind(key, object, size, os.time()):step()
+    end)
+    if not ok and err then
+        err = err:gsub("\n.*", "") -- remove stacktrace
+        logger.err("CacheSQLite:insert() failed:", err)
+    end
+
+    self.current_size = self.current_size + size
+    self:closeDB()
+    return ok, size
+end
+
+--- Retrieves an object if it is in the cache and updates its access time.
+--- @param key string
+--- @return any
+function CacheSQLite:check(key)
+    self:openDB()
+    -- Update access time and retrieve value
+    local stmt = self.db:prepare([[
+        UPDATE cache SET last_access = ? WHERE key = ?
+        RETURNING value;
+    ]])
+    local row = stmt:reset():bind(os.time(), key):step()
```

Comment:
(Comment valid across all the `os.time` calls)

`time.now` instead?

Or even `UIManager:getTime()` if a coarse, cached value from our last UI frame is sufficient?

Reply:
I like that os.time() is human-legible when looking at the table, fwiw. `UIManager:getTime()` probably makes the most sense as an alternative, or in any case I don't see what higher granularity could be desired for.

![Screenshot_20250125_222535](https://github.com/user-attachments/assets/b281a7b0-1baa-4f71-8af1-dc6f4d5829e8)